pkg/etcdutil/etcdutil.go

```diff
@@ -149,7 +157,15 @@ func DefragmentData(defragCtx context.Context, client *clientv3.Client, endpoint
 }
 
 // TakeAndSaveFullSnapshot takes full snapshot and save it to store
-func TakeAndSaveFullSnapshot(ctx context.Context, client *clientv3.Client, store brtypes.SnapStore, lastRevision int64, cc *compressor.CompressionConfig, suffix string, logger *logrus.Entry) (*brtypes.Snapshot, error) {
+func TakeAndSaveFullSnapshot(ctx context.Context, clientFactory client.Factory, store brtypes.SnapStore, lastRevision int64, cc *compressor.CompressionConfig, suffix string, logger *logrus.Entry) (*brtypes.Snapshot, error) {
```

Comment:
Why not pass `client.MaintenanceCloser` here instead of `client.Factory` here? That would be more consistent with other places in the PR. WDYT?

Reply:
Hi @amshuman-kr , thanks for review. Hope you are doing well. 
`TakeAndSaveFullSnapshot`  is getting called from [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/compactor/compactor.go#L150)(in compaction) and [here](https://github.com/gardener/etcd-backup-restore/blob/5a7fd60f0bf704672a987dcab83c0c156041ff21/pkg/snapshot/snapshotter/snapshotter.go#L303)(in snapshotter).
In compaction, yes it make sense to pass  `client.MaintenanceCloser` instead of `client.Factory` as we already have  `client.MaintenanceCloser` object.
But In snapshotter,  to pass  `client.MaintenanceCloser` we have to first get the `client.MaintenanceCloser` object in [takeFullSnapshot func call](https://github.com/gardener/etcd-backup-restore/blob/5a7fd60f0bf704672a987dcab83c0c156041ff21/pkg/snapshot/snapshotter/snapshotter.go#L264).
So, I thought it will be better to make  `TakeAndSaveFullSnapshot` more abstract by letting the `TakeAndSaveFullSnapshot` to create client by itself and we have to just pass the `client.Factory`

Reply:
`takeFullSnapshot` is already constructing the client object in [the existing code](https://github.com/gardener/etcd-backup-restore/blob/5a7fd60f0bf704672a987dcab83c0c156041ff21/pkg/snapshot/snapshotter/snapshotter.go#L269-L275). So, why not keep the same flow even the changes in this PR?

Reply:
yeah, make sense but don't you think it will be overhead for caller to create client instead of just passing the configuration and let the `TakeAndSaveFullSnapshot` do all work

Reply:
> but don't you think it will be overhead for caller to create client instead of just passing the configuration and let the TakeAndSaveFullSnapshot do all work

If it is overhead it is already overhead. Also, in the compaction case, it would lead to unnecessary creation of an additional client object. If there were more places like the `takeFullSnapshot` then it might have tilted the argument in favour of creating a client. But as it stands, creating a client outside and passing it to `TakeAndSaveFullSnapshot` seems better.

Reply:
https://github.com/gardener/etcd-backup-restore/pull/381#discussion_r712076941, so I'm not moving `clientMaintenance` to inside the  `opts.NeedDefragmentation {}` block because of scope as now I need `clientMaintenance` to pass `TakeAndSaveFullSnapshot`